File: BinarySearchTreeIterator/BinarySearchTreeIterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    BSTIterator(TreeNode *root) {
        TreeNode* node = root;
        while(node != nullptr) {
        	nodes.push(node);
        	node = node->left;
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !nodes.empty();
    }

    /** @return the next smallest number */
    int next() {
        TreeNode* node = nodes.top();
        nodes.pop();
        TreeNode* newNode = node->right;
    	while(newNode != nullptr) {
    		nodes.push(newNode);
    		newNode = newNode->left;
    	}
        return node->val;
    }
private:
	stack<TreeNode*> nodes;
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: BinarySearchTreeIterator/BinarySearchTreeIterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.nodes = []
        node = root
        while node is not None:
            self.nodes.append(node)
            node = node.left

    def hasNext(self):
        """
        :rtype: bool
        """
        return True if len(self.nodes) != 0 else False

    def next(self):
        """
        :rtype: int
        """
        node = self.nodes.pop()
        newNode = node.right
        while newNode is not None:
            self.nodes.append(newNode)
            newNode = newNode.left
        return node.val

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())